### PR TITLE
Add support of FCLASS.S

### PR DESCRIPTION
**Pull request by 2011eric**

### CODE REVIEW

**Review by jserv — CHANGES_REQUESTED** [review]
Amend the git commit message to render the results of RISC-V Architectural Tests.

File: riscv_private.h [hunk 1]
```diff
@@ -322,13 +323,13 @@ static inline uint32_t calc_fclass(uint32_t f) {
   /* 0x020    rs1 is positive subnormal */
   out |= (!expn && frac && !sign) ? 0x020 : 0;
   /* 0x040    rs1 is positive normal */
-  out |= (expn && expn < 0x78000000 && !sign) ? 0x040 : 0;
+  out |= (expn && (expn != FMASK_EXPN) && !sign) ? 0x040 : 0;
   /* 0x080    rs1 is +INF */
-  out |= (f == 0x7f800000) ? 0x080 : 0;
+  out |= (expn == FMASK_EXPN && !frac && !sign) ? 0x080 : 0;
```
Comment (jserv) [inline riscv_private.h:328]: Shall we update the description of these comments?

Comment (2011eric) [reply riscv_private.h:328]: How should I update the description?

**Comment by 2011eric** [comment]
Print the input and output of fclass:

```
static inline uint32_t calc_fclass(uint32_t f) {
  const uint32_t sign = f & FMASK_SIGN;
  const uint32_t expn = f & FMASK_EXPN;
  const uint32_t frac = f & FMASK_FRAC;

  /* TODO: optimize with a binary decision tree */

  uint32_t out = 0;
  /* 0x001    rs1 is -INF */
  out |= (f == 0xff800000) ? 0x001 : 0;
  /* 0x002    rs1 is negative normal */
  out |= (expn && (expn != FMASK_EXPN) && sign) ? 0x002 : 0;
  /* 0x004    rs1 is negative subnormal */
  out |= (!expn && frac && sign) ? 0x004 : 0;
  /* 0x008    rs1 is -0 */
  out |= (f == 0x80000000) ? 0x008 : 0;
  /* 0x010    rs1 is +0 */
  out |= (f == 0x00000000) ? 0x010 : 0;
  /* 0x020    rs1 is positive subnormal */
  out |= (!expn && frac && !sign) ? 0x020 : 0;
  /* 0x040    rs1 is positive normal */
  out |= (expn && (expn != FMASK_EXPN) && !sign) ? 0x040 : 0;
  /* 0x080    rs1 is +INF */
  out |= (expn == FMASK_EXPN && !frac && !sign) ? 0x080 : 0;
  /* 0x100    rs1 is a signaling NaN */
  out |= (expn == FMASK_EXPN && frac && !(frac & FMASK_QNAN)) ? 0x100 : 0;
  /* 0x200    rs1 is a quiet NaN */
  out |= (expn == FMASK_EXPN && (frac & FMASK_QNAN))? 0x200 : 0;

  printf("%08x ", f);
  printf("%08x\n", out);
  
  return out;
}
```

```
00000000 00000010
bf800000 00000002
3f800000 00000040
ffaaaaaa 00000100
7f800001 00000100
ffc55555 00000200
7fc00001 00000200
ffc00000 00000200
7fc00000 00000200
ff800000 00000001
7f800000 00000080
ff7fffff 00000002
7f7fffff 00000040
80855555 00000002
00800001 00000040
80800000 00000002
00800000 00000040
807fffff 00000004
007fffff 00000020
807ffffe 00000004
00000002 00000020
80000001 00000004
00000001 00000020
80000000 00000008
00000000 00000010
00000000 00000010
00000000 00000010
00000000 00000010
00000000 00000010
00000000 00000010
00000000 00000010
00000000 00000010
80000000 00000008
inferior exit code 0
```

File: riscv_private.h (hunk 1 above)
Comment (jserv) [reply riscv_private.h:328]: Explicitly mentioning signedness checks would be better.

**Comment by jserv** [comment]
> Print the input and output of fclass:

Instead of dumping the difference in the signature, you should rewrite down the descriptive text in human-readable form. That is, you can summarize the riscv-arch-test reports.

**Comment by jserv** [comment]
I amended the commit message, and further issues should be confirmed later.

**Pull request merged**